operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1728 @@
+
+		pool, found := ipam.poolStore.GetByName(lbRange.originPool)
+		if !found {
+			ipam.logger.Warnf("Bad state detected, store contains lbRange for pool '%s' but missing the pool", lbRange.originPool)
+			continue
+		}
+
+		// If there is no selector, all services match
+		if pool.Spec.ServiceSelector != nil {
+			selector, err := slim_meta_v1.LabelSelectorAsSelector(pool.Spec.ServiceSelector)
+			if err != nil {
+				return nil, nil, false, fmt.Errorf("Making selector from pool '%s' label selector: %w", pool.Name, err)
+			}
+
+			if !selector.Matches(sv.Labels) {
+				continue
+			}
+		}
+
+		// Attempt to allocate the next IP from this range.
+		newIp, err := lbRange.allocRange.AllocateNext()
+		if err != nil {
+			// If the range is full, mark it.
+			if errors.Is(err, ipallocator.ErrFull) {
+				full = true
+				continue
+			}
+
+			ipam.logger.WithError(err).Error("Allocate next IP from lb range")
+			continue
+		}
+
+		return &newIp, lbRange, false, nil
+	}
+
+	return nil, nil, full, nil
+}
+
+// serviceIPFamilyRequest checks which families of IP addresses are requested
+func (ipam *LBIPAM) serviceIPFamilyRequest(svc *core_v1.Service) (IPv4Requested, IPv6Requested bool) {
+	if svc.Spec.IPFamilyPolicy != nil {
+		switch *svc.Spec.IPFamilyPolicy {
+		case core_v1.IPFamilyPolicySingleStack:
+			if len(svc.Spec.IPFamilies) > 0 {
+				if svc.Spec.IPFamilies[0] == core_v1.IPFamily(IPv4Family) {
+					IPv4Requested = true
+				} else {
+					IPv6Requested = true
+				}
+			} else {
+				if ipam.ipv4Enabled {
+					IPv4Requested = true
+				} else if ipam.ipv6Enabled {
+					IPv6Requested = true
+				}
+			}
+
+		case core_v1.IPFamilyPolicyPreferDualStack:
+			if len(svc.Spec.IPFamilies) > 0 {
+				for _, family := range svc.Spec.IPFamilies {
+					if family == core_v1.IPFamily(IPv4Family) {
+						IPv4Requested = ipam.ipv4Enabled
+					}
+					if family == core_v1.IPFamily(IPv6Family) {
+						IPv6Requested = ipam.ipv6Enabled
+					}
+				}
+			} else {
+				// If no IPFamilies are specified
+
+				IPv4Requested = ipam.ipv4Enabled
+				IPv6Requested = ipam.ipv6Enabled
+			}
+
+		case core_v1.IPFamilyPolicyRequireDualStack:
+			IPv4Requested = ipam.ipv4Enabled
+			IPv6Requested = ipam.ipv6Enabled
+		}
+	} else {
+		if len(svc.Spec.IPFamilies) > 0 {
+			if svc.Spec.IPFamilies[0] == core_v1.IPFamily(IPv4Family) {
+				IPv4Requested = true
+			} else {
+				IPv6Requested = true
+			}
+		} else {
+			if ipam.ipv4Enabled {
+				IPv4Requested = true
+			} else if ipam.ipv6Enabled {
+				IPv6Requested = true
+			}
+		}
+	}
+
+	return IPv4Requested, IPv6Requested
+}
+
+// Handle the addition of a new IPPool
+func (ipam *LBIPAM) handleNewPool(pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+	// Sanity check that we do not yet know about this pool.
+	if _, found := ipam.poolStore.GetByName(pool.GetName()); found {
+		ipam.logger.Warnf("LB IPPool with uid '%s' has been created, but a LB IP Pool with the same uid already exists", pool.GetUID())
```

Comment:
nit: log name, not uid here. UIDs are opaque and non-selectable.

Reply:
And use `WithField` and friends.